probe/src/flash/flasher.rs

```diff
@@ -0,0 +1,583 @@
+
+        let mut data = vec![0; algo.instructions.len()];
+        self.session.probe.read_block32(algo.load_address, &mut data)?;
+
+        assert_eq!(&algo.instructions, &data.as_slice());
+        log::debug!("RAM contents match flashing algo blob.");
+
+        log::debug!("Preparing Flasher for region:");
+        log::debug!("{:#?}", &self.region);
+        log::debug!("Double buffering enabled: {}", self.double_buffering_supported);
+        let mut flasher = ActiveFlasher {
+            session: self.session,
+            region: self.region,
+            double_buffering_supported: self.double_buffering_supported,
+            _operation: core::marker::PhantomData,
+        };
+
+        // Execute init routine if one is present.
+        if let Some(pc_init) = algo.pc_init {
+            log::debug!("Running init routine.");
+            let result = flasher.call_function_and_wait(
+                pc_init,
+                address,
+                clock.or(Some(0)),
+                Some(O::operation()),
+                None,
+                true
+            )?;
+
+            if result != 0 {
+                return Err(FlasherError::Init(result));
+            }
+        }
+
+        Ok(flasher)
+    }
+
+    pub fn run_erase<T, E: From<FlasherError>>(
+        &mut self,
+        f: impl FnOnce(&mut ActiveFlasher<Erase>) -> Result<T, E> + Sized
+    ) -> Result<T, E> {
+        // TODO: Fix those values (None, None).
+        let mut active = self.init(None, None)?;
+        let r = f(&mut active)?;
+        active.uninit()?;
+        Ok(r)
+    }
+
+    pub fn run_program<T, E: From<FlasherError>>(
+        &mut self,
+        f: impl FnOnce(&mut ActiveFlasher<Program>) -> Result<T, E> + Sized
+    ) -> Result<T, E> {
+        // TODO: Fix those values (None, None).
+        let mut active = self.init(None, None)?;
+        let r = f(&mut active)?;
+        active.uninit()?;
+        Ok(r)
+    }
+
+    pub fn run_verify<T, E: From<FlasherError>>(
+        &mut self,
+        f: impl FnOnce(&mut ActiveFlasher<Verify>) -> Result<T, E> + Sized
+    ) -> Result<T, E> {
+        // TODO: Fix those values (None, None).
+        let mut active = self.init(None, None)?;
+        let r = f(&mut active)?;
+        active.uninit()?;
+        Ok(r)
+    }
+
+    pub fn flash_block(
+        mut self,
+        address: u32,
+        data: &[u8],
+        chip_erase: Option<bool>,
+        smart_flash: bool,
+        fast_verify: bool,
+    ) -> Result<(), FlasherError> {
+        if !self.region.range.contains_range(&(address..address + data.len() as u32)) {
+            return Err(FlasherError::AddressNotInRegion(address, self.region.clone()));
+        }
+
+        let mut fb = FlashBuilder::new(self.region.range.start);
+        fb.add_data(address, data).expect("Add Data failed");
+        fb.program(self, chip_erase, smart_flash, fast_verify, true).expect("Add Data failed");
```

Comment:
Maybe return an error here,  we already have a function which returns a Result.

Reply:
Error conversion is not possible as of now. I need to do quite some refactoring of the flash algo bits where I will do this :) Created an issue to track it.